src/sentry/sentry_metrics/querying/metadata.py

```diff
@@ -0,0 +1,185 @@
+from datetime import datetime
+from typing import Generator, List, Optional, Sequence, Set, TypedDict
+
+from sentry.api.utils import InvalidParams
+from sentry.models.organization import Organization
+from sentry.models.project import Project
+from sentry.sentry_metrics.querying.utils import get_redis_client_for_ingest
+from sentry.utils import json, metrics
+
+DAY_IN_SECONDS = 86400
+
+
+class CodeLocationQuery(TypedDict):
+    organization_id: int
+    project_id: int
+    metric_mri: str
+    timestamp: int
+
+
+class CodeLocationPayload(TypedDict):
+    function: Optional[str]
+    module: Optional[str]
+    filename: Optional[str]
+    abs_path: Optional[str]
+    lineno: Optional[int]
+
+
+class MetricCodeLocations(TypedDict):
+    query: CodeLocationQuery
+    code_locations: Sequence[CodeLocationPayload]
+
+
+def _get_day_timestamps(
+    start: datetime,
+    end: datetime,
+) -> Set[int]:
+    # We compute the approximate day timestamps, assuming UTC as the timezone.
+    start_in_seconds = int(start.timestamp() / DAY_IN_SECONDS) * DAY_IN_SECONDS
+    end_in_seconds = int(end.timestamp() / DAY_IN_SECONDS) * DAY_IN_SECONDS
+
+    current_time = start_in_seconds
+    timestamps = set()
+    while current_time <= end_in_seconds:
+        timestamps.add(current_time)
+        current_time += DAY_IN_SECONDS
+
+    return timestamps
+
+
+def get_cache_key_for_code_location(
+    organization_id: int, project_id: int, metric_mri: str, timestamp: int
+) -> str:
+    # TODO: implement mri conversion.
+    encoded_mri = metric_mri
+    return f"mm:l:{{{organization_id}}}:{project_id}:{encoded_mri}:{timestamp}"
+
+
+class CodeLocationsFetcher:
+    # The maximum number of keys that can be fetched by the fetcher.
+    #
+    # The estimation was naively done by supposing at most 10 metrics with 2 projects and at most 90 timestamps.
+    MAXIMUM_KEYS = 2000
+    # The size of the batch of keys that are fetched by endpoint.
+    #
+    # Batching is done via Redis pipeline and the goal is to improve the performance of the system.
+    BATCH_SIZE = 50
+
+    def __init__(
+        self,
+        organization: Organization,
+        projects: Set[Project],
+        metric_mris: Set[str],
+        timestamps: Set[int],
+    ):
+        self._organization = organization
+        self._projects = projects
+        self._metric_mris = metric_mris
+        self._timestamps = timestamps
+
+        self._redis_client = get_redis_client_for_ingest()
+
+        self._validate()
+
+    def _validate(self):
+        total_combinations = len(self._projects) * len(self._metric_mris) * len(self._timestamps)
```

Comment:
can we track the number of total combinations with a metric here, to get a sense the amount of requests we are making

Reply:
I was thinking about that, could be a nice use case for DDM but I didn't do it because such requests are made by the frontend which is controlled by us, thus we would not get many useful insights. WDYT?

Reply:
Yeah but we currently cant use DDM in the frontend as the sdk is not there yet :D